### DISCUSSION TITLE
Question about thinning process in v.voronoi

### DISCUSSION THREAD

**Discussion by g-marques-git** [OSGeo/grass#2623] [Q&A]
Hi,

I'm trying to implement the thinning process of v.voronoi for skeleton in another tool (python).

The thinning process take place in skeleton.c : https://github.com/OSGeo/grass/blob/releasebranch_7_8/vector/v.voronoi/skeleton.c from the line 273.

My understanding is it is only based on length of lines and multiple call to v.clean: rmdangle algorithm with incrementing threshold should do the same (as stated in v.clean documentation). But in the source code there is a morphof parameter that i have trouble understanding. Could someone enlighten me about his use please ?

Regards

**Comment by wenzeslaus** [comment]
Hi, here is a rough code for calling _v.voronoi_ from Python code:

```python
import grass.script as gs
import grass.script.setup  # for <=8.2

# Assuming we have created /some/path/nc_spm_08/PERMANENT and imported or linked the data.
with gs.setup.init("/some/path/nc_spm_08/PERMANENT"):
    gs.run_command("v.voronoi", input="urbanarea", output="urbanarea_skeleton", flags="s", thin=2000)
```

If that does not work for you, can please give us more information on what you are doing?

**Reply by g-marques-git** [reply]
Thanks for the reply.

However running grass tools in python is not my problem. I'm trying to understand what the morphof argument in the source code stands for. To try to replicate the method in a tool i'm building for thinning skeleton obtained by other methods. In my understanding length and morphof are used to remove dangle after constructing the skeleton. But i do not grasp the use of morphof in this process.

Regards

**Reply by wenzeslaus** [reply]
Why are you replicating the method instead of using the one in GRASS GIS? In the community, there are likely people willing to help you apply GRASS GIS to your use case. If you need to modify the method, that's possible, too, through a pull request.

**Reply by g-marques-git** [reply]
I'm not really replicating. 
v.voronoi.skeleton have numerical instability degrading results in some case. I need another method to offer to users if this happen on a dataset. For now i use r.thin and apply v.clean:rmdangle with an incremental value for the length threshold. The result are good but the thinning process in v.voronoi seems way better. So i was wondering if only the length was used in v.voronoi.skeleton. As i was browsing the source code i noticed the morphof argument and couldn't see what it does. I was wondering if the difference could come from here.
